cmd/app/options/options.go

```diff
@@ -0,0 +1,186 @@
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	cmversioned "github.com/jetstack/cert-manager/pkg/client/clientset/versioned"
+	cmclient "github.com/jetstack/cert-manager/pkg/client/clientset/versioned/typed/certmanager/v1"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"istio.io/istio/pkg/jwt"
+	"istio.io/istio/pkg/spiffe"
+	"istio.io/istio/security/pkg/server/ca/authenticate"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/openstack"
+	cliflag "k8s.io/component-base/cli/flag"
+)
+
+// Options is a struct to hold options for cert-manager-istio-agent
+type Options struct {
+	*AppOptions
+	*CertManagerOptions
+	*TLSOptions
+	*KubeOptions
+}
+
+type AppOptions struct {
+	logLevel string
+	Logr     *logrus.Entry
+}
+
+type CertManagerOptions struct {
+	issuerName  string
+	issuerKind  string
+	issuerGroup string
+
+	Namespace   string
+	PreserveCRs bool
+	IssuerRef   cmmeta.ObjectReference
+}
+
+type TLSOptions struct {
+	RootCACertFile        string
+	ServingAddress        string
+	ServingCertificateTTL time.Duration
+}
+
+type KubeOptions struct {
+	kubeConfigFlags *genericclioptions.ConfigFlags
+
+	CMClient   cmclient.CertificateRequestInterface
+	KubeClient kubernetes.Interface
+	Auther     authenticate.Authenticator
+}
+
+func New() *Options {
+	return &Options{
+		AppOptions:         new(AppOptions),
+		CertManagerOptions: new(CertManagerOptions),
+		TLSOptions:         new(TLSOptions),
+		KubeOptions:        new(KubeOptions),
+	}
+}
+
+func (o *Options) Prepare(cmd *cobra.Command) *Options {
+	o.addFlags(cmd)
+	return o
+}
+
+func (o *Options) Complete() error {
+	logLevel, err := logrus.ParseLevel(o.logLevel)
+	if err != nil {
+		return fmt.Errorf("failed to parse --log-level %q: %s",
+			o.logLevel, err)
+	}
+
+	logr := logrus.New()
+	logr.SetOutput(os.Stdout)
+	logr.SetLevel(logLevel)
+	o.Logr = logrus.NewEntry(logr)
+
+	restConfig, err := o.kubeConfigFlags.ToRESTConfig()
+	if err != nil {
+		return fmt.Errorf("failed to build kubernetes rest config: %s", err)
+	}
+
+	o.KubeClient, err = kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return fmt.Errorf("failed to build kubernetes client: %s", err)
+	}
+
+	o.Auther = authenticate.NewKubeJWTAuthenticator(o.KubeClient, "Kubernetes", nil, spiffe.GetTrustDomain(), jwt.PolicyThirdParty)
+
+	cmClient, err := cmversioned.NewForConfig(restConfig)
+	if err != nil {
+		return fmt.Errorf("failed to build cert-manager client: %s", err)
+	}
+
+	o.CMClient = cmClient.CertmanagerV1().CertificateRequests(o.Namespace)
+
+	o.IssuerRef = cmmeta.ObjectReference{
+		Name:  o.issuerName,
+		Kind:  o.issuerKind,
+		Group: o.issuerGroup,
+	}
+
+	return nil
+}
+
+func (o *Options) addFlags(cmd *cobra.Command) {
+	var nfs cliflag.NamedFlagSets
```

Comment:
This is a way of flags i haven't seen before? Is this something specific?

Reply:
It's basically used for pretty printing flags into names groupings;

```bash
Usage:
  cert-manager-istio-agent [flags]

App flags:

  -v, --log-level string   Log level (debug, info, warn, error, fatal, panic). (default "info")

TLS flags:

      --root-ca-cert string                File location of a PEM encoded Root CA certificate to be used as root of trust for TLS. If empty, the CA returned from the cert-manager issuer will be used.
  -a, --serving-address string             Address to serve certificates gRPC service. (default "0.0.0.0:443")
  -t, --serving-certificate-ttl duration   TTL duration of serving certificates. Will be renewed after 2/3 of the duration. (default 24h0m0s)

Cert-manager flags:

  -c, --certificate-namespace string    Namespace to request certificates. (default "istio-system")
  -g, --issuer-group string             Group of the issuer to sign istio workload certificates. (default "cert-manager.io")
  -k, --issuer-kind string              Kind of the issuer to sign istio workload certificates. (default "Issuer")
  -u, --issuer-name string              Name of the issuer to sign istio workload certificates. (default "istio-ca")
  -d, --preserve-certificate-requests   If enabled, will preserve created CertificateRequests, rather than deleting when they are ready.

Kubernetes flags:

      --as string                      Username to impersonate for the operation
      --as-group stringArray           Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
....................
```